### ISSUE TITLE
Remove ConstructProjectionUsing 

### ISSUE THREAD

**Issue by lbargaoanu** [AutoMapper/AutoMapper#2795]
Replace it with expression based ConstructUsing. It seems like a good time, as we've already bumped the version. But I guess it's debatable. Why wouldn't the same apply to ProjectUsing?

**Comment by jbogard** [comment]
The general argument against is that not every ConstructUsing can be represented as an expression.

Another option would be to remove the ConstructProjectionUsing (really any expression-specific method) and instead have an overload:

```c#
        IMappingExpression<TSource, TDestination> ConstructUsing(Expression<Func<TSource, TDestination>> ctor);
        IMappingExpression<TSource, TDestination> ConstructUsing(Func<TSource, ResolutionContext, TDestination> ctor);
```

If you want to use a complex `Func`, you need to use the more complex overload. I'd like to go one step further and remove `ResolveUsing` but those have an additional complication in that their results are mapped.

**Comment by lbargaoanu** [comment]
Yes, but I was trying to say, ProjectUsing vs ConvertUsing, isn't it the same?

**Comment by jbogard** [comment]
Ah, yes. Those two as well.

**Comment by jbogard** [comment]
So would we remove the "ResolveUsing" overloads that just map via `Func` to collapse to `MapFrom`? Save `ResolveUsing` for actual resolvers?

**Comment by lbargaoanu** [comment]
You mean MapFrom(Func), I suppose (not MapFrom(Expression)). That makes sense, but I think it's too big of a change. I guess we could provide some built in extension methods, but that would defeat the purpose as most people would probably use those.

**Comment by lbargaoanu** [comment]
Updated with ResolveUsing.

**Comment by lbargaoanu** [comment]
But the problem here is that we might break a lot of  projections when going from Func to Expression. What used to apply only for Map will apply for ProjectTo too now.

**Comment by jbogard** [comment]
I'm not suggesting breaking `ProjectTo`, but this will break some `Func` based versions of the overloads. Mainly, if you're using a `Func`-based version that's getting replaced with `Expression`, you *must* use expressions now. Whether or not that expression can be used by `ProjectTo` is different.

What this will explicitly prevent though is people wanting to use different configurations for `ProjectTo` and `Map`, which is fine by me.

If you have a complex lambda statement that can't be converted to an expression, you'll just need to use the overloads that allow `Func`.

**Comment by jbogard** [comment]
Also, I'm fine with the `ResolveUsing` changes. That makes sense - those overloads use an object rather than func.

**Comment by lbargaoanu** [comment]
Yes, but the end result is some code used to work, you upgrade AM, it compiles, and it no longer works for ProjectTo because the older version didn't use it (still works for Map). Right? :) So I'm not sure of the impact here.

**Comment by jbogard** [comment]
I don't think so - the old version you'd have to use `ConstructProjectionUsing` for example, and that method would be gone. It'd only affect people that used both methods. Which I don't think we'd want to support, separate config for different mapping execution strategies.

If you were using both, you now have to choose. I'm OK with that.

**Comment by lbargaoanu** [comment]
I have ConstructUsing(Func). That one becomes CostructUsing(Expression) after the upgrade. Let's say it compiles. The new version will use it for ProjectTo, but EF doesn't support that. The same applies to ConvertUsing. I think :)

**Comment by jbogard** [comment]
It's possible - which would only break people that supplied *both* `ConstructUsing` and `ConstructProjectionUsing`. For those folks, which is weird, they'd need to adjust their `ConstructUsing` to use the more complex `Func`-based overload.

It won't break anyone at runtime today. Only break at compile-time. Anyone working at runtime today should continue to work.

If they had `ConvertUsing(Func)` and it now works with expression, *then* decided to use EF, it could break. But that's true with `ConvertProjectionUsing` today - you could pass in a legal expression but illegal EF expression.

So the impact is twofold, I think:

- It is a breaking compile change for those using `Expression`-based methods
- Lambda-statement based `Func` that now use `Expression` will need to switch overloads to the more complex versions that accept more params

**Comment by lbargaoanu** [comment]
I'm not proposing anything different :) I'm just trying to think through the consequences. Clearly we're breaking things. But I guess the alternative is to leave it as it is.

**Comment by jbogard** [comment]
Some more thoughts on the `MapFrom`/`ResolveUsing`, it's a different pattern than the `ConvertUsing` paradigm. Even the interface name is `Resolve`, right now the proposal has:

- Expression, Func, and class-based type mapping with `ConvertUsing`
- Expression and Func-based construction using `ConstructUsing`
- Expression and Func-based member mapping using `MapFrom`
- Class-based member mapping using `ResolveUsing`
- Func and class-based mapping actions using `BeforeMap` and `AfterMap`

So for consistency sake, I'd prefer to consolidate all member mapping to `MapFrom` like all the others.

**Comment by jbogard** [comment]
Fixed by #2809, #2807, and #2808
